File: julien-lectures/CODE/Program_7_8.m
function [t,X,Y] = Program_7_8(n,tau,gamma,Y0,N0,MaxTime)
%
% 
%
% Program_7_8( n, tau, gamma, Y0, N0, MaxTime)
%      This is the MATLAB version of program 7.8 from page 285 of 
% "Modeling Infectious Disease in humans and animals" 
% by Keeling & Rohani.
% 
% It is the pairwise approximation to the SIS model on a random network of
% N individuals, each with n contacts.
% 
% Initially it is assumed that there is no correlation between infected and
% susceptible individuals in the network. 
%

% Sets up default parameters if necessary.
if nargin == 0
   n=4;
   tau=0.1;
   gamma=0.05;
   Y0=1;
   N0=10000;
   MaxTime=100;
end
X0=N0-Y0;

% Checks all the parameters are valid
CheckGreater(tau,0,'beta');
CheckGreaterOrEqual(gamma,0,'gamma');
CheckGreater(n,0,'nu');
CheckGreater(N0,0,'N0');
CheckGreater(Y0,0,'Y0');
CheckGreaterOrEqual(X0,0,'X0=N0-Y0');

X=X0; XY=n*Y0*X0/N0; 

% The main iteration 
options = odeset('RelTol', 1e-4);
[t, pop]=ode45(@Diff_7_8,[0 MaxTime],[X XY] ,options,n, tau, gamma, N0);

X=pop(:,1); XY=pop(:,2); Y=N0-X;

% plots the graphs 
subplot(3,1,1)
h=plot(t,Y,'-g');
xlabel 'Time';
ylabel 'Infectious'

subplot(3,1,2) 
h=plot(t,XY,'-r');
xlabel 'Time';
ylabel '[XY] pairs'

% this last graph is the effective reduction in transmission due to the
% negative spatial correlations between susceptibles and infecteds.
subplot(3,1,3) 
h=plot(t,XY./(n*X.*Y/N0),'-r');
xlabel 'Time';
ylabel 'relative XY correlation'




% Calculates the differential rates used in the integration.
function dPop=Diff_7_8(t,pop, n, tau, gamma, N)

X=pop(1); XY=pop(2);

dPop=zeros(2,1);

% Taken from Box 7.4

dPop(1)= gamma*(N-X) - tau*XY;
dPop(2)= tau*(n-1)*(n*X-XY)*XY/(n*X) + gamma*(n*N-n*X-XY) - tau*XY - tau*(n-1)*XY*XY/(n*X) - gamma*XY;


% Does a simple check on the value
function []=CheckGreaterOrEqual(Parameter, Value, str)

if Parameter<Value
    error('Parameter %s (=%g) is less than %g',str,Parameter,Value);
end

% Does a simple check on the value
function []=CheckGreater(Parameter, Value, str)

if Parameter<=Value
    error('Parameter %s (=%g) is less than %g',str,Parameter,Value);
end

File: julien-lectures/CODE/Program_7_8.py
#!/usr/bin/env python

####################################################################
###    This is the PYTHON version of program 7.8 from page 285 of  #
### "Modeling Infectious Disease in humans and animals"            #
### by Keeling & Rohani.										   #
###																   #
### It is the pairwise approximation to the SIS model on a random  #
### network of N individuals, each with n contacts.                #
###																   #
### Initially it is assumed that there is no correlation between   #
### infected and susceptible individuals in the network.           #
####################################################################

###################################
### Written by Ilias Soumpasis    #
### ilias.soumpasis@ucd.ie (work) #
### ilias.soumpasis@gmail.com	  #
###################################

import scipy.integrate as spi
import numpy as np
import pylab as pl

n=4;
tau=0.1;
gamma=0.05;
Y0=1.0;
N0=10000;
ND=MaxTime=100.0;
TS=1.0
X0=N0-Y0
XY0=n*Y0*X0/N0; 

INPUT=np.hstack((X0,XY0))

def diff_eqs(INP,t):  
	'''The main set of equations'''
	Y=np.zeros((2))
	V = INP   
	Y[0] = gamma*(N0-V[0]) - tau*V[1];
	Y[1] = tau*(n-1)*(n*V[0]-V[1])*V[1]/(n*V[0]) + gamma*(n*N0-n*V[0]-V[1]) - tau*V[1]\
	- tau*(n-1)*V[1]*V[1]/(n*V[0]) - gamma*V[1];
	return Y   # For odeint

t_start = 0.0; t_end = ND; t_inc = TS
t_range = np.arange(t_start, t_end+t_inc, t_inc)
RES = spi.odeint(diff_eqs,INPUT,t_range)
X = RES[:,0]
XY= RES[:,1]
Y=N0-X


#Ploting
pl.subplot(311)
pl.plot(Y, '-g')
pl.ylabel('Infectious')
pl.subplot(312)
pl.plot(XY, '-r')
pl.ylabel('[XY] pairs')
pl.subplot(313)
pl.plot(XY/((n*X)*(Y/N0)), '-r')
pl.ylabel('relative XY correlation')
pl.xlabel('Time (Years)')
print XY/(n*X*Y/N0)

pl.show()
